cmd/containerd-shim-runhcs-v1/service_internal.go

```diff
@@ -447,12 +447,13 @@ func (s *service) shutdownInternal(ctx context.Context, req *task.ShutdownReques
 		return empty, nil
 	}
 
-	if req.Now {
-		os.Exit(0)
-	}
-	// TODO: JTERRY75 if we dont use `now` issue a Shutdown to the ttrpc
-	// connection to drain any active requests.
-	os.Exit(0)
+	s.shutOnce.Do(func() {
+		// TODO: should taskOrPod be deleted/set to nil?
+		// TODO: is there any extra leftovers of the shimTask/Pod to clean? ie: verify all handles are closed?
+		s.GracefulShutdown = !req.Now
+		close(s.shut)
+	})
+
```

Comment:
Since you linked the runc v2 handling of this method, should we also bail if the shim is still hosting containers? https://github.com/containerd/containerd/blob/77d53d2d230c3bcd3f02e6f493019a72905c875b/runtime/v2/runc/task/service.go#L463-L466

Reply:
So, from what I understand, the runc code does not explicitly track an init/pod container, and it uses `len(s.containers)` to track if all the containers have been deleted. 
We don't properly delete tasks from the pods internal data store ([yet](https://github.com/microsoft/hcsshim/pull/1271)), so we don't have a way of knowing if we are still servicing pods. The closest analogy would be to check if the sandbox container is still executing, but that would rely on the assumption that, in between stopping the sandbox container and the shutdown command for it, containerd (or whomever) deleted all the containers properly.

What do you think is the best way forward?
I like the idea of, on deleting the sandbox container (if there is one) or the init task, we zero`service.taskOrPod` to signal that everything was deleted, and use that as the sentinal.

Reply:
If we can't do anything here yet until that other change lands then I'm fine with doing nothing. Your idea for zeroing taskOrPod sounds fine afterwards, but obviously my feelings may change once seeing it in action 😆